src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs

```diff
@@ -225,18 +226,43 @@ private static uint HandleEventConnected(State state, ref ConnectionEvent connec
         {
             if (!state.Connected)
             {
-                // Connected will already be true for connections accepted from a listener.
-                Debug.Assert(!Monitor.IsEntered(state));
-                SOCKADDR_INET inetAddress = MsQuicParameterHelpers.GetINetParam(MsQuicApi.Api, state.Handle, QUIC_PARAM_LEVEL.CONNECTION, (uint)QUIC_PARAM_CONN.LOCAL_ADDRESS);
+                if (state.IsServer)
+                {
+                    state.Connected = true;
+                    MsQuicListener.State? listenerState = state.ListenerState;
+                    state.ListenerState = null;
 
-                Debug.Assert(state.Connection != null);
-                state.Connection._localEndPoint = MsQuicAddressHelpers.INetToIPEndPoint(ref inetAddress);
-                state.Connection.SetNegotiatedAlpn(connectionEvent.Data.Connected.NegotiatedAlpn, connectionEvent.Data.Connected.NegotiatedAlpnLength);
-                state.Connection = null;
+                    if (listenerState != null)
+                    {
+                        if (listenerState.PendingConnections.TryRemove(state.Handle.DangerousGetHandle(), out MsQuicConnection? connection))
+                        {
+                            // Move connection from pending to Accept queue and hand it out.
+                            if (listenerState.AcceptConnectionQueue.Writer.TryWrite(connection))
+                            {
+                                return MsQuicStatusCodes.Success;
+                            }
+                            // Listener is closed
+                            connection.Dispose();
+                        }
+                    }
 
-                state.Connected = true;
-                state.ConnectTcs!.SetResult(MsQuicStatusCodes.Success);
-                state.ConnectTcs = null;
+                    return MsQuicStatusCodes.UserCanceled;
```

Comment:
I assume this part corresponds to the PR description:
> I updated the code to simple return failure instead of throwing. (that will clean the connection)

Could you explain exactly what "clean the connection" means in this context? We have methods like Shutdown (finish using) and Close (free memory). Does it mean we don't need to call one/both of them? Won't it lead to double free since we're calling `connection.Dispose()` few lines above?

I'm just trying to understand what this exactly mean, I'm not saying it's wrong.

Reply:
From what I know, when the event callback returns something else than success code, MsQuic will notify peer, cleanup the connection internally and we will get "SHUTDOWN_COMPLETED" event. That should release the GC handle and that will start final description of the connection object.  
We may or may not call the Dispose,  depending how we will get to the end. If we do that should close the safe handle and tell MsQuic we are done. Attempt to make call on the connection would cause ODE from the SafeHandle.
That does not mean MsQuic would immediately free it. AFAIK they grab references when they need to and the object should be actually freed when squid is all done. 
cc: @nibanks to confirm

Reply:
So it substitutes Shutdown but not Close (free), which we still need to call sometime later.